### ISSUE TITLE
Problem with filtering and ordering on related data types

### ISSUE THREAD

**Issue by osterkraft** [graphql-nexus/nexus-plugin-prisma#345]
I'm having problems with adding filtering/ordering in a case like this:

```
export const ContentItem = objectType({
  name: "ContentItem",
  definition(t) {
    t.model.id();
    t.model.name({
      filtering: true
    });
  }
});
```

I expect to be able to make queries like this one:

```
query($lanuage: String) {
  findManyContentItem{
    id
    name(where: {language: $language}) {
      value
      language
    }
  }
}
```

But I get the following type of error:

```
Using ts-node version 8.3.0, typescript version 3.5.2
Error: Could not find filtering argument for ContentItem.name
    at NexusPrismaBuilder.argsForQueryOrModelField (node_modules/@generated/nexus-prisma/index.js:444:23)
    at NexusPrismaBuilder.computeArgsFromField (node_modules/@generated/nexus-prisma/index.js:434:25)
    at Object.acc.(anonymous function).opts [as name] (node_modules/@generated/nexus-prisma/index.js:568:139)
    at Object.definition (src/index.ts:219:13)
    at SchemaBuilder.walkOutputType (node_modules/@prisma/nexus/src/builder.ts:1192:9)
    at SchemaBuilder.walkTypes (node_modules/@prisma/nexus/src/builder.ts:456:16)
    at SchemaBuilder.getFinalTypeMap (node_modules/@prisma/nexus/src/builder.ts:464:10)
    at buildTypes (node_modules/@prisma/nexus/src/builder.ts:1313:18)
    at makeSchemaInternal (node_modules/@prisma/nexus/src/builder.ts:1357:9)
    at Object.makeSchema (node_modules/@prisma/nexus/src/builder.ts:1408:9)
[ERROR] 22:44:52 Error: Could not find filtering argument for ContentItem.name
```

**Data model (simplified):**

```
model ContentItem {
  id        String           @default(cuid()) @id @unique
  name      StringItem[]     @relation(name: "RelatedStringItemContent")
}

model StringItem {
  id                  String       @default(cuid()) @id @unique
  value               String
  createdAt           DateTime     @default(now())
  key                 String?
  language            String?
  contentItem        ContentItem? @relation(name: "RelatedStringItemContent")
}
```

Adding filtering/ordering to definitions where a field references another document of the same type seems to work, but when there is a link between related data types it stops working. The arguments are accepted by TypeScript so I assume it should be working.

Running Nexus/Prisma2 (tried up to prisma2@2.0.0-preview-6).

**Comment by iherger** [comment]
I am having similar problems.

The following GraphQL query (exposed through nexus/prisma2) does not work:
```graphql
{
  registrations(where: { event: { id: "cjzbjh9gi00061e9ktor940hb" } }) {
    id
    user {
      firstname
      familyname
    }
  }
}
```
The error message is:
```
Field "id" is not defined by type QueryFindManyRegistrationWhereInput.
```
which indicates that nexus-prisma is expecting the wrong type for `event: {}`. In my mind, the correct type should be `QueryFindManyEventWhereInput`.

The corresponding photon query works fine:
``` typescript
  const regsWithUser = await photon.registrations.findMany({
    include: { user: true },
    where: { event: { id: 'cjzbjh9gi00061e9ktor940hb' } },
  });
```

Having the ability to filter and order by related data types is very important for many queries (except the most basic ones), so I would be very happy if this could be fixed soon.

**Comment by 5achinJani** [comment]
https://github.com/prisma/nexus/issues/202
https://github.com/prisma/nexus/issues/203

**Comment by jasonkuhrt** [comment]
Trying to figure out if this is a bug or feature (@Weakky ?).

I only see `alias` as available option for scaler-types on model exposer functions e.g.:

![image](https://user-images.githubusercontent.com/284476/64196123-2205f780-ce51-11e9-8c88-9e7cb78de20d.png)

vs options on relational model exposers:

![image](https://user-images.githubusercontent.com/284476/64196175-42ce4d00-ce51-11e9-9990-ff12105ebc3e.png)

**Comment by iherger** [comment]
I would think that this is a bug, not a feature...

Reasons: 
1. you can filter on related models in photon
2. it also worked with prisma1/nexus/nexus-prisma

**Comment by Weakky** [comment]
That was indeed a bug, as `name` isn't a scalar on @iherger's example. However, I just tried locally and it works fine. Have you tried upgrading to `nexus-prism@0.0.1-beta.7` ?

**Comment by iherger** [comment]
@Weakky , can you tell me what works fine for you locally? I am using `beta.7` and I am not able to filter on related models (it works with photon, but not with nexus/nexus-prisma).

**Comment by Weakky** [comment]
I took your prisma schema, your graphql types, and the filtering worked fine. Meaning the server started without error, and I could introspect the graphql endpoint and see the `where` argument on `ContentItem.name(..., where: ...)`

```ts
export const ContentItem = objectType({
  name: 'ContentItem',
  definition(t) {
    t.model.name({ filtering: true })
  },
})

export const StringItem = objectType({
  name: 'StringItem',
  definition(t) {
    t.model.id()
  },
})

const Query = objectType({
  name: 'Query',
  definition(t) {
    t.crud.stringitems()
  },
})
```

**Comment by iherger** [comment]
Thanks, @Weakky. This was @osterkraft's model, and it seems to be working indeed now, which is great news!

However, my problem above (https://github.com/prisma/nexus-prisma/issues/345#issuecomment-521535197) still remains, I think. It's very similar, but not the same. If it's easier, I can open another issue for this.

 If I find the time, I will set up an example with all the different filters on different levels.

**Comment by iherger** [comment]
Coming back to this.

I have set up a small repo with three cases that you can look at (one works [for strange reasons, in my mind] and two don't work).

Repo: https://github.com/iherger/prisma2-relationfilters

Consider the following three queries:

1.
```graphql
{
  posts(where: { author: { id: { equals: "ck057djn40000yq9knoke7nte" } } }) {
    title
    published
    author {
      id
      name
    }
  }
}
```

This works for me.

2.
```graphql
{
  posts(where: { author: { name: { equals: "Bob" } } }) {
    title
    published
    author {
      id
      name
    }
  }
}
```

This fails with the error:
```
Field "name" is not defined by type QueryFindManyPostWhereInput.
```
There seems to be some recursion of the expected types going on, see screenshot:
<img width="1323" alt="Screenshot 2019-09-04 at 18 33 54" src="https://user-images.githubusercontent.com/36133712/64274397-99b44f00-cf43-11e9-8476-94e0647c09e5.png">

3.
```graphql
{
  users(where: { posts: { some: { published: { equals: true } } } }) {
    name
    posts {
      title
      published
    }
  }
}
```

This fails with the error:

```
Field "published" is not defined by type QueryFindManyUserWhereInput.
```

**Comment by Weakky** [comment]
Thanks @iherger for your reproduction. I could easily target the problem here.

@jasonkuhrt As a side-effect of the filtering and ordering arguments customization (eg: `t.crud.posts({ filtering: { id: true, author: true } })`), nexus-prisma needs to come up with inputs names scoped to their parent type + field name

eg: By default, given a `User` output type, we might have a `UserWhereInput` input type.
But because we can customize it everywhere, nexus-prisma aliases it to `QueryUsersWhereInput` for such a schema 

```graphql
type Query {
  users(where: QueryUsersWhereInput): [User!]!
}
```
(type name: `Query` + field name: `users` = `QueryUsers`)

The issue occurs for some of the fields of `QueryUsersWhereInput`.
- For the `AND` and `OR` fields, we can reference the same type `QueryUsersWhereInput`, so no issue.
- For the `String/Int/BooleanFilter` types, there is no issue either as nexus-prisma only allow to customize input types on one level of depth.
- The problem lies with relational filters inside `QueryUsersWhereInput` 👇

```graphql
type Query {
  users(where: QueryUsersWhereInput): [User!]!
}

input QueryUsersWhereInput {
  id: StringFilter
  AND: [QueryUsersWhereInput]
  OR: [QueryUsersWhereInput]
  post: WhatShouldBeTheNameOfThisType?? <--
}
```

At the moment, `QueryUsersWhereInput.post` is of type `QueryUsersWhereInput` too, which is the cause of the problem. It should be of type `PostWhereInput`, but I'm not sure yet whether we should alias these types as well or not.

Because we can only customize on one level of depth, the type of the field `QueryUsersWhereInput.post` won't ever be customized anyway. However, having input type's names scoped to their parent type name + field name for top-level field arguments, but not for the fields of input types makes the input type's names less predictable.

On the other hand, aliasing the types of fields of input types (eg: `QueryUsersWhereInput.post`) will create a lot more clutter/confusion in the final schema

We need to make a decision or find another approach to solve the problem.

PS: Let's fix #396 while solving this issue

**Comment by Weakky** [comment]
Actually, this is a duplicate of #331. I'll close this one in favor of #331 because the main issue of that issue is already solved.
